SLING-8932 - Redesign DistributionSubscriber to offload responsibilities

cschneider opened the pull request.

tmaret commented on src/main/java/org/apache/sling/distribution/journal/impl/subscriber/BookKeeper.java:
```diff
@@ -0,0 +1,300 @@
+        context.stop();
+    }
+
+    private void store(ResourceResolver resolver, PackageStatus packageStatus) throws PersistenceException {
```
I think we should keep `storeStatus` because it effectively stores the status. Suggest to s/store/storeStatus back.

cschneider replied:
Make sense

tmaret commented on src/main/java/org/apache/sling/distribution/journal/impl/subscriber/BookKeeper.java:
```diff
@@ -0,0 +1,300 @@
+        context.stop();
+    }
+
+    public void sendStoredStatus() throws InterruptedException, IOException {
```
I think that the previous `sendStoredStatus` implementation and flow was cleaner and clearer. Would you be ok getting that part back ?

cschneider replied:
I agree.. I will use the old logic.

tmaret commented on src/main/java/org/apache/sling/distribution/journal/impl/subscriber/CommandPoller.java:
```diff
@@ -0,0 +1,105 @@
+
+    public CommandPoller(MessagingProvider messagingProvider, Topics topics, String subSlingId, String subAgentName, boolean editable) {
+        this.subSlingId = subSlingId;
+        // TODO Auto-generated constructor stub
```
The TODO seems resolved

cschneider replied:
Ok

cschneider commented on src/main/java/org/apache/sling/distribution/journal/impl/subscriber/BookKeeper.java:
```diff
@@ -0,0 +1,300 @@
+        distributionMetricsService.getPackageDistributedDuration().update((currentTimeMillis() - createdTime), TimeUnit.MILLISECONDS);
+        packageRetries.clear(pkgMsg.getPubAgentName());
+
+        Event event = DistributionEvent.eventImporterImported(pkgMsg, subAgentName);
```
Makes sense. I did not take into account that a commit might fail.

cschneider commented on src/main/java/org/apache/sling/distribution/journal/impl/subscriber/BookKeeper.java:
```diff
@@ -0,0 +1,300 @@
+            store(importerResolver, new PackageStatus(IMPORTED, offset, pkgMsg.getPubAgentName()));
+        }
+        storeOffset(importerResolver, offset);
+        distributionMetricsService.getImportedPackageSize().update(pkgMsg.getPkgLength());
```
Ok .. will change it.

cschneider commented on src/main/java/org/apache/sling/distribution/journal/impl/subscriber/DistributionSubscriber.java:
```diff
@@ -405,61 +302,61 @@ private DistributionResponse executeUnsupported(DistributionRequest request) {
             // and then process it
             try (Timer.Context context = distributionMetricsService.getProcessQueueItemDuration().time()) {
                 processQueueItem(item);
+                queueItemsBuffer.remove();
```
I introduced an error here. We should only remove the item and change the metrics in case of no IllegalStateException. I will fix this

cschneider replied:
Done in next commit.

cschneider commented:
I fixed the issues @tmaret found in the review. To minimize the interactions with BookKeeper I moved the whole retry and metrics logic there. I also found and fixed an additional issue introduced during the initial commit.

tmaret commented:
I don't have time to do a throughout review again, but the changes LGTM. 

It makes much more sense to keep the three repository write ops and single commit at the same place rather than spreading this logic all over the place.
